file_store/src/price_tracker.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{Error, FileStore, FileType, Result, Settings};
+use chrono::{Duration, Utc};
+use futures::stream::{StreamExt, TryStreamExt};
+use helium_proto::{services::price_oracle::PriceOracleReportV1, BlockchainTokenTypeV1, Message};
+use std::{collections::HashMap, sync::Arc};
+
+use tokio::sync::RwLock;
+type Prices = Arc<RwLock<HashMap<BlockchainTokenTypeV1, u64>>>;
+
+pub struct PriceTracker {
+    prices: Prices,
+}
+
+impl PriceTracker {
+    pub async fn start(
+        settings: &Settings,
+        shutdown: triggered::Listener,
+    ) -> Result<(Self, impl std::future::Future<Output = Result>)> {
+        let prices = Arc::new(RwLock::new(HashMap::new()));
+
+        let prices_clone = prices.clone();
+        let settings_clone = settings.clone();
+        let handle = tokio::spawn(async move { run(settings_clone, prices_clone, shutdown).await });
+
+        Ok((PriceTracker { prices }, async move {
+            match handle.await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(err)) => Err(err),
+                Err(err) => Err(Error::from(err)),
+            }
+        }))
+    }
+
+    pub async fn price(&self, token_type: BlockchainTokenTypeV1) -> Result<u64> {
+        self.prices
+            .read()
+            .await
+            .get(&token_type)
+            .cloned()
+            .ok_or(Error::PriceNotAvailable())
+    }
+}
+
+async fn run(settings: Settings, prices: Prices, shutdown: triggered::Listener) -> Result {
+    let file_store = FileStore::from_settings(&settings).await?;
+    let mut trigger = tokio::time::interval(std::time::Duration::from_secs(30));
+    let mut after = Utc::now() - Duration::minutes(10);
+
+    loop {
+        let shutdown = shutdown.clone();
+
+        tokio::select! {
+            _ = shutdown => {
+                tracing::info!("PriceTracker shutting down");
+                break;
+            }
+            _ = trigger.tick() => {
+                let files = file_store.list_all(FileType::PriceReport, after, None).await?;
+
+                for file in files {
+                    tracing::debug!("processing pricing report file {}", file.key);
+                    after = file.timestamp;
+                    file_store.stream_file(file)
```

Comment:
did we drop the "source unordered" from the file streams so these will always stream in oldest to newest and this mutating timestamp will be accurately moving forward with each file?

Reply:
This does not process files out of order or concurrently so we will always end up with latest prices